taichi/program/state_flow_graph.cpp

```diff
@@ -983,55 +947,16 @@ bool StateFlowGraph::demote_activation() {
     if (nodes.size() <= 1)
       continue;
 
-    auto snode = nodes[0]->meta->snode;
-
-    auto list_state = AsyncState(snode, AsyncState::Type::list);
-
-    TI_ASSERT(snode != nullptr);
-
-    std::unique_ptr<IRNode> new_ir = nodes[0]->rec.ir_handle.clone();
-
-    OffloadedStmt *offload = new_ir->as<OffloadedStmt>();
-    Block *body = offload->body.get();
-
-    // TODO: for now we only deal with the top level. Is there an easy way to
-    // extend this part?
-    auto consts = ConstExprPropagation::run(body, [](Stmt *stmt) {
-      if (stmt->is<ConstStmt>()) {
-        return true;
-      } else if (stmt->is<LoopIndexStmt>())
-        return true;
-      return false;
-    });
-
-    bool demoted = false;
-    for (int k = 0; k < (int)body->statements.size(); k++) {
-      Stmt *stmt = body->statements[k].get();
-      if (auto ptr = stmt->cast<GlobalPtrStmt>(); ptr && ptr->activate) {
-        bool can_demote = true;
-        // TODO: test input mask?
-        for (auto ind : ptr->indices) {
-          if (consts.find(ind) == consts.end()) {
-            // non-constant index
-            can_demote = false;
-          }
-        }
-        if (can_demote) {
-          modified = true;
-          ptr->activate = false;
-          demoted = true;
-        }
-      }
-    }
-    // TODO: cache this part
-    auto new_handle = IRHandle(new_ir.get(), ir_bank_->get_hash(new_ir.get()));
-    ir_bank_->insert(std::move(new_ir), new_handle.hash());
-    auto new_meta = get_task_meta(ir_bank_, nodes[0]->rec);
-    if (demoted) {
-      for (int j = 1; j < (int)nodes.size(); j++) {
+    auto new_handle = ir_bank_->demote_activation(nodes[0]->rec.ir_handle);
+    if (new_handle != nodes[0]->rec.ir_handle) {
+      modified = true;
+      nodes[1]->rec.ir_handle = new_handle;
+      nodes[1]->meta = get_task_meta(ir_bank_, nodes[1]->rec);
+      for (int j = 2; j < (int)nodes.size(); j++) {
         nodes[j]->rec.ir_handle = new_handle;
-        nodes[j]->meta = new_meta;
+        nodes[j]->meta = nodes[1]->meta;
       }
+      // TODO: do we need to break here?
```

Comment:
```suggestion
      // For every "demote_activation" call, we only optimize for 
      // a single key in std::map<std::pair<IRHandle, Node *>, std::vector<Node *>> tasks
      // since the graph probably needs to be rebuild after demoting part of the tasks.
```

Reply:
Hope that explains why we `break`. Not breaking may work as well, but breaking seems more trivially correct.